src/vm/instruction.ts

```diff
@@ -10,68 +33,303 @@
-enum RegisterFlag {
+export enum RegisterFlag {
   ApRegisterFlag = 0,
   FpRegisterFlag = 1,
 }
-type DstRegister = RegisterFlag;
-type Op0Register = RegisterFlag;
 
 // Op1Src
-enum Op1Src {
+export enum Op1Src {
   Imm = 0,
   FpPlusOffOp1 = 1,
   ApPlusOffOp1 = 2,
   Op0 = 4,
 }
 
 // ResLogic
-enum ResLogic {
+export enum ResLogic {
   Op1 = 0,
   Add = 1,
   Mul = 2,
   Unconstrained = 3,
 }
 
 // Pc Update
-enum PcUpdate {
+export enum PcUpdate {
   PcUpdateRegular = 0,
   PcUpdateJump = 1,
   PcUpdateJumpRel = 2,
   PcUpdateJnz = 3,
 }
 
 // Ap update
-enum ApUpdate {
+export enum ApUpdate {
   ApUpdateRegular = 0,
   ApUpdateAdd = 1,
-  AppUpdateAdd1 = 2,
-  AppUpdateAdd2 = 3,
+  ApUpdateAdd1 = 2,
+  ApUpdateAdd2 = 3,
 }
 
 // Fp Update
-enum FpUpdate {
+export enum FpUpdate {
   FpUpdateRegular = 0,
   FpUpdateApPlus2 = 1,
   FpUpdateDst = 2,
 }
 
-enum Opcode {
+export enum Opcode {
   NoOp = 0,
-  AssertEq = 1,
-  Call = 2,
-  Ret = 4,
+  Call = 1,
+  Ret = 2,
+  AssertEq = 4,
 }
 
-export type Instruction = {};
+export type Instruction = {
+  off0: Int16;
+  off1: Int16;
+  off2: Int16;
+  dstReg: RegisterFlag;
+  op0Reg: RegisterFlag;
+  op1Addr: Op1Src;
+  resLogic: ResLogic;
+  pcUpdate: PcUpdate;
+  apUpdate: ApUpdate;
+  fpUpdate: FpUpdate;
+  opcode: Opcode;
+};
+
+export function decodeInstruction(
+  encodedInstruction: Uint64
+): Result<Instruction, VMError> {
+  const highBit = 1n << 63n;
+  const dstRegMask = UnsignedInteger.toUint16(0x01).unwrap(); // safe to unwrap
+  const dstRegOff = UnsignedInteger.toUint16(0).unwrap(); // safe to unwrap
+  const op0RegMask = UnsignedInteger.toUint16(0x02).unwrap(); // safe to unwrap
+  const op0RegOff = UnsignedInteger.toUint16(1).unwrap(); // safe to unwrap
+  const op1SrcMask = UnsignedInteger.toUint16(0x1c).unwrap(); // safe to unwrap
+  const op1SrcOff = UnsignedInteger.toUint16(2).unwrap(); // safe to unwrap
+  const resLogicMask = UnsignedInteger.toUint16(0x60).unwrap(); // safe to unwrap
+  const resLogicOff = UnsignedInteger.toUint16(5).unwrap(); // safe to unwrap
+  const pcUpdateMask = UnsignedInteger.toUint16(0x0380).unwrap(); // safe to unwrap
+  const pcUpdateOff = UnsignedInteger.toUint16(7).unwrap(); // safe to unwrap
+  const apUpdateMask = UnsignedInteger.toUint16(0x0c00).unwrap(); // safe to unwrap
+  const apUpdateOff = UnsignedInteger.toUint16(10).unwrap(); // safe to unwrap
+  const opcodeMask = UnsignedInteger.toUint16(0x7000).unwrap(); // safe to unwrap
+  const opcodeOff = UnsignedInteger.toUint16(12).unwrap(); // safe to unwrap
+
+  if ((highBit & encodedInstruction) !== 0n) {
+    return new Err(HighBitSetError);
+  }
+
+  const mask = UnsignedInteger.toUint64(0xffffn).unwrap(); // safe to unwrap
+  const shift = UnsignedInteger.toUint64(16n).unwrap(); // safe to unwrap
+
+  // Get the offset by masking and shifting the encoded instruction
+  const offsetDstUint64 = UnsignedInteger.uint64And(encodedInstruction, mask);
```

Comment:
ERRATUM: this is nonsense, please ignore
should be `&` and it's bigint, not number,
and it doesn't work because 
```
const offsetDstUint64 = encodedInstruction & mask
```
is not typed as Uint64